plugins/inputs/solr/solr.go

```diff
@@ -0,0 +1,424 @@
+			"errors":                      metrics.Stats.Errors,
+			"expunge_deletes":             metrics.Stats.ExpungeDeletes,
+			"optimizes":                   metrics.Stats.Optimizes,
+			"rollbacks":                   metrics.Stats.Rollbacks,
+			"soft_autocommits":            metrics.Stats.SoftAutocommits,
+		}
+		acc.AddFields(
+			fmt.Sprintf("solr_%s", strings.ToLower(category)),
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func gatherCacheMetrics(mbeansJSON json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]Cache
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJSON, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		coreFields := map[string]interface{}{
+			"class_name":           metrics.Class,
+			"cumulative_evictions": metrics.Stats.CumulativeEvictions,
+			"cumulative_hitratio":  metrics.Stats.CumulativeHitratio,
+			"cumulative_hits":      metrics.Stats.CumulativeHits,
+			"cumulative_inserts":   metrics.Stats.CumulativeInserts,
+			"cumulative_lookups":   metrics.Stats.CumulativeLookups,
+			"evictions":            metrics.Stats.Evictions,
+			"hitratio":             metrics.Stats.Hitratio,
+			"hits":                 metrics.Stats.Hits,
+			"inserts":              metrics.Stats.Inserts,
+			"lookups":              metrics.Stats.Lookups,
+			"size":                 metrics.Stats.Size,
+			"warmup_time":          metrics.Stats.WarmupTime,
+		}
+		acc.AddFields(
+			fmt.Sprintf("solr_%s", strings.ToLower(category)),
+			coreFields,
+			map[string]string{
+				"core":    core,
+				"handler": name},
+			measurementTime,
+		)
+	}
+	return nil
+}
+
+func (s *Solr) gatherCoreStats(url string, core string, acc telegraf.Accumulator) error {
+
+	for _, category := range []string{"CORE", "QUERYHANDLER", "UPDATEHANDLER", "CACHE"} {
+		metrics := &Metrics{}
+		if err := s.gatherData(fmt.Sprintf("%s&cat=%s", url, category), metrics); err != nil {
+			acc.AddError(fmt.Errorf("JSON fetch error: %s", err))
+		}
+		switch category {
+		case "CORE":
+			if err := gatherCoreMetrics(metrics.SolrMbeans[1], core, category, acc); err != nil {
+				acc.AddError(fmt.Errorf("core category: %s", err))
+			}
+		case "QUERYHANDLER":
+			if err := gatherQueryHandlerMetrics(metrics.SolrMbeans[1], core, category, acc); err != nil {
+				acc.AddError(fmt.Errorf("query handler category: %s", err))
+			}
+		case "UPDATEHANDLER":
+			if err := gatherUpdateHandlerMetrics(metrics.SolrMbeans[1], core, category, acc); err != nil {
+				acc.AddError(fmt.Errorf("update handler category: %s", err))
+			}
+		case "CACHE":
+			if err := gatherCacheMetrics(metrics.SolrMbeans[1], core, category, acc); err != nil {
+				acc.AddError(fmt.Errorf("cache category: %s", err))
+			}
+		default:
+			err := errors.New("unrecognized category")
+			acc.AddError(fmt.Errorf("category: %s", err))
```

Comment:
These 2 lines should be merged or at least handled differently. Line 397 creates `err`, but then 398 converts `err` into a string, losing any reason for creating an error object.